### PR TITLE
bug1118003 - Transparency Panel

### PR DESCRIPTION
**Pull request by fabi1cazenave**
https://bugzilla.mozilla.org/show_bug.cgi?id=1118003

### CODE REVIEW

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=f5127c132413)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=2b570dcfcefb)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=2a1dcfa1d7d1)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=92635319ec2e)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=30df8efe9e00)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=aa967565247d)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=ed8a60a68601)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=4d0ca5071a57)

File: apps/privacy-panel/test/unit/app_list_test.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+  test('should get an array of app groups (one group per vendor)',
+    function(done) {
+      assert.deepEqual(this.subject.getSortedApps('vendor'), {
+        'Flying Platypus': [ fakeApp1 ],
+        'Mighty Duck': [ fakeApp2, fakeApp3 ]
+      });
+      done();
+    }
+  );
```
Comment (etiennesegonzac) [inline apps/privacy-panel/test/unit/app_list_test.js]: Looks like we're missing tests for the `app.trust` property

File: apps/privacy-panel/js/app_list.js [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+  }
+
+  /**
+   * Add properties to the {DOMApplication} instances in the _application list.
+   *   .name:        localized name
+   *   .trust:       trust level (= certified, privileged, web)
+   *   .vendor:      developer name
+   *   .iconURL:     URL of the best icon for the current display
+   *   .permissions: filtered list of permissions that are actually used
+   */
```
Comment (etiennesegonzac) [inline apps/privacy-panel/js/app_list.js]: How about making this a `_makeAppRepresentation` returning new objects with just the properties we need.
Or, alternatively, a clone of non-functions properties + our custom ones.

(yes, what prompted this bikeshed is the fact that this is called in the middle of a very functional chain but is then having `defineProperty`-level side effect ;))

**Comment by etiennesegonzac** [comment]
You'll be quicker than myself to debug since it comes form the panel black magic but on every visibility change I get multiple `TypeError: event.detail is undefined"`

**Comment by etiennesegonzac** [comment]
We'll need to go further with the integrations tests (or quadruple down on unit-tests but I'd vote for integrations tests).

The integration framework lets us inject apps in the profile before running the test [1].
This will enable to cover much more cases without having our tests fail every time an app is added/removed from gaia.

Here's the wish list:
* inject an app asking for the contacts permission and asserting its presence on the permissions->contacts panel
* inject 2 apps with specific us to cover the alphabetical sort of the Applications panel
* same one with 2 specific vendors
* same one with 2 specific trust level

[1] https://github.com/mozilla-b2g/gaia/blob/master/apps/system/test/marionette/task_manager_test.js#L24

File: apps/privacy-panel/test/unit/app_list_test.js (hunk 1 above)
Comment (fabi1cazenave) [inline apps/privacy-panel/test/unit/app_list_test.js]: The test just above this one sorts apps by `app.trust`, just like this test sorts apps by `app.vendor`. Please clarify if you think these tests aren’t meaningful.

File: apps/privacy-panel/js/app_list.js (hunk 2 above)
Comment (fabi1cazenave) [inline apps/privacy-panel/js/app_list.js]: Good point, thanks for the suggestion. Bikeshedding is always good anyway. :-)

**Comment by fabi1cazenave** [comment]
> You'll be quicker than myself to debug since it comes form the panel black magic but on every visibility change I get multiple `TypeError: event.detail is undefined"`

Ugh, I don’t get those. Everything works fine on my Keon with a recent GP flash (20150107013010).
How do you trigger the visibility changes? (I do edge swipes and home button presses)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=3da650fb70cf)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=79a8f51bc090)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=7812e8a1bc25)

**Comment by fabi1cazenave** [comment]
@etiennesegonzac your wishlist has been happily implemented. However, it looks like `navigator.mozPermissionSettings` is not supported for the fake apps in these Marionette tests, so I’ve used a “real” app (Browser) to test the permission display.

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=a11976b82131)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=d406ddee2044)

File: apps/privacy-panel/locales/permissions.en-US.properties [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+perm-browser-description = Enables the app to implement a browser in an iframe.
+perm-camera-description = Take photos, shoot video, record audio, and control the camera.
+
+perm-contacts-description = Add, read, or modify contacts from the address book on the device and read contacts from the SIM.
+perm-contacts-readonly-description = Read contacts from the address book on the device and read contacts from the SIM.
+perm-contacts-readwrite-description = Read or modify contacts from the address book on the device and read contacts from the SIM.
+perm-contacts-readcreate-description = Add, read, or modify contacts from the address book on the device and read contacts from the SIM.
+perm-contacts-createonly-description = Add contacts to the address book on the device.
+
+perm-desktop-notification-description = Display a notification on the user’s desktop.
```
Comment (flodolo) [inline apps/privacy-panel/locales/permissions.en-US.properties]: Do we actually need/want a string about desktop notifications on Gaia?

File: apps/privacy-panel/locales/permissions.en-US.properties [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+perm-device-storage-videos-description = Add, read or modify video files stored on the device.
+perm-device-storage-videos-readonly-description = Read video files stored on the device.
+perm-device-storage-videos-readwrite-description = Read or modify video files stored on the device.
+perm-device-storage-videos-readcreate-description = Add, read or modify video files stored on the device.
+perm-device-storage-videos-createonly-description = Store new video files on the device.
+
+perm-fmradio-description = Control the FM radio.
+perm-geolocation-description = Obtain the current location of the user.
+perm-input-description = Allows the app to act as a virtual keyboard by listening to focus change events in other apps.
```
Comment (flodolo) [inline apps/privacy-panel/locales/permissions.en-US.properties]: For consistency I'd say `Allow the app…`

File: apps/privacy-panel/locales/permissions.en-US.properties [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+perm-storage-description = Utilize storage without size limitations.
+perm-systemXHR-description = Allows anonymous cross-origin XHR without the target site having CORS enabled.
+perm-tcp-socket-description = Create TCP sockets and communicate over them.
+perm-video-capture-description = Access video input devices, e.g. camera.
+
+# Implicit permissions (certified/internal applications)
+perm-attention-description = Allow content to open a window in front of all other content.
+perm-audio-channel-ringer-description = Play sounds for incoming phone calls.
+perm-audio-channel-telephony-description = Play sounds for phone calls, VoIP calls.
+perm-audio-channel-publicnotification-description = Forced camera shutter sounds.
```
Comment (flodolo) [inline apps/privacy-panel/locales/permissions.en-US.properties]: I wonder what this actually means/does

File: apps/privacy-panel/locales/permissions.en-US.properties [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+perm-power-description = Turn the screen on or off, control CPU, device power, etc.
+
+perm-settings-description = Configure or read device settings.
+perm-settings-readonly-description = Read device settings.
+perm-settings-readwrite-description = Read or modify device settings.
+perm-settings-readcreate-description = Add, read or modify device settings.
+perm-settings-createonly-description = Add device settings.
+
+perm-sms-description = Send and receive SMS messages.
+perm-telephony-description = Access all telephony-related APIs to make and recieve phone calls.
```
Comment (flodolo) [inline apps/privacy-panel/locales/permissions.en-US.properties]: typo: receive

File: apps/privacy-panel/locales/permissions.en-US.properties [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+perm-time-description = Set current time.
+perm-voicemail-description = Access voicemail.
+perm-webapps-manage-description = Manage installed Open Web Apps.
+perm-wifi-manage-description = Enumerate available WiFi networks, get signal strength, connect to a network.
```
Comment (flodolo) [inline apps/privacy-panel/locales/permissions.en-US.properties]: WiFi -> Wi-Fi

File: apps/privacy-panel/locales/permissions.en-US.properties (hunk 4 above)
Comment (fabi1cazenave) [inline apps/privacy-panel/locales/permissions.en-US.properties]: Good catch, thanks.

File: apps/privacy-panel/locales/permissions.en-US.properties (hunk 3 above)
Comment (fabi1cazenave) [inline apps/privacy-panel/locales/permissions.en-US.properties]: That’s… a very good point. :-)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=f04f29c38b9e)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=e23d2d6666e8)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=405c3a20b030)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=d34ba899288c)

File: apps/privacy-panel/test/marionette/tc_main_test.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+var assert = require('assert');
+var TcMainPanel = require('./lib/panels/tc_main');
+
+/**
+ * For this test, we're using these three fake apps:
+ *
+ * +--------+-------------------+----------------------+-----------------------+
+ * |        | mightyD           | chubbyU              | flyingP               |
+ * +--------+-------------------+----------------------+-----------------------+
+ * | name   | Mighty Duck       | Chubby Unicorn       | Flying Platypus       |
```
Comment (etiennesegonzac) [inline apps/privacy-panel/test/marionette/tc_main_test.js]: Well that's pretty much an auto-r+, but I'll read the rest anyway :)

File: apps/privacy-panel/js/app_list.js [hunk 9]
```diff
@@ -0,0 +1,321 @@
+/**
+ * App List module.
+ *
+ * @module AppList
+ * @return {Object}
+ */
+define([], function() {
+  'use strict';
+
+  var _ = navigator.mozL10n.get;
```
Comment (zbraniecki) [inline apps/privacy-panel/js/app_list.js]: no need for this

File: apps/privacy-panel/test/unit/app_list_test.js [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+      'mocks/mock_navigator_moz_apps',
+      'mocks/mock_l10n'
+    ],
+    function(mozApps, mozL10n) {
+      realMozApps = navigator.mozApps;
+      navigator.mozApps = mozApps;
+      navigator.mozApps.mApps = apps;
+
+      realMozL10n = navigator.mozL10n;
+      navigator.mozL10n = mozL10n;
```
Comment (zbraniecki) [inline apps/privacy-panel/test/unit/app_list_test.js]: do you need this?

File: apps/privacy-panel/js/tc/applications.js [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    },
+
+    _showAppSeparator: function _showAppSeparator(separator, l10nPrefix) {
+      if (!separator) {
+        return;
+      }
+      var header = document.createElement('header');
+      var title = document.createElement('h2');
+      if (l10nPrefix) {
+        title.setAttribute('data-l10n-id', l10nPrefix + separator);
```
Comment (zbraniecki) [inline apps/privacy-panel/js/tc/applications.js]: Do we have all the potential l10nPrefix + separator pairs in .properties file? Is there a risk that we will have a missing entity here?

File: apps/privacy-panel/js/app_list.js (hunk 9 above)
Comment (fabi1cazenave) [inline apps/privacy-panel/js/app_list.js]: This is used to localize the permissions (line 154-155), in order to sort them by name.

File: apps/privacy-panel/js/tc/applications.js (hunk 11 above)
Comment (fabi1cazenave) [inline apps/privacy-panel/js/tc/applications.js]: Yes, we have all the l10nPrefix + separator pairs in the l10n file. The prefix is always `tc-trust-`, and the separator is either `certified`, `privileged` or `web`.

File: apps/privacy-panel/test/unit/app_list_test.js (hunk 10 above)
Comment (fabi1cazenave) [inline apps/privacy-panel/test/unit/app_list_test.js]: yes, because `js/app_list.js` uses mozL10n to sort permissions by localized name.

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=c37906ba18ca)

**Comment by try-server-hook** [comment]
[fabi1cazenave Fabien Cazenave (fabi1cazenave)](https://github.com/fabi1cazenave) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=bf636d7b4d2e)

**Pull request merged**